onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -1,193 +1,231 @@
 }
 
 template <typename DstType>
-void CastFromStringData(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  const int64_t len = shape.Size();
-  if (std::is_same<DstType, float>::value) {
-    auto* mutable_data = out.MutableData<float>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stof(in.Data<std::string>()[i]);
-    }
-  } else if (std::is_same<DstType, double>::value) {
-    auto* mutable_data = out.MutableData<double>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stod(in.Data<std::string>()[i]);
-    }
-  } else if (std::is_same<DstType, int8_t>::value) {
-    auto* mutable_data = out.MutableData<int8_t>();
-    for (int i = 0; i < len; ++i) {
-      int temp_i = std::stoi(in.Data<std::string>()[i]);
-      mutable_data[i] = static_cast<int8_t>(temp_i);
-    }
-  } else if (std::is_same<DstType, uint8_t>::value) {
-    auto* mutable_data = out.MutableData<uint8_t>();
-    for (int i = 0; i < len; ++i) {
-      unsigned long temp_ui = std::stoul(in.Data<std::string>()[i]);
-      mutable_data[i] = static_cast<uint8_t>(temp_ui);
-    }
-  } else if (std::is_same<DstType, int16_t>::value) {
-    auto* mutable_data = out.MutableData<int16_t>();
-    for (int i = 0; i < len; ++i) {
-      int temp_i = std::stoi(in.Data<std::string>()[i]);
-      mutable_data[i] = static_cast<int16_t>(temp_i);
-    }
-  } else if (std::is_same<DstType, uint16_t>::value) {
-    auto* mutable_data = out.MutableData<uint16_t>();
-    for (int i = 0; i < len; ++i) {
-      unsigned long temp_ui = std::stoul(in.Data<std::string>()[i]);
-      mutable_data[i] = static_cast<uint16_t>(temp_ui);
-    }
-  } else if (std::is_same<DstType, int32_t>::value) {
-    auto* mutable_data = out.MutableData<int32_t>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stol(in.Data<std::string>()[i]);
-    }
-  } else if (std::is_same<DstType, uint32_t>::value) {
-    auto* mutable_data = out.MutableData<uint32_t>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stoul(in.Data<std::string>()[i]);
-    }
-  } else if (std::is_same<DstType, int64_t>::value) {
-    auto* mutable_data = out.MutableData<int64_t>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stoll(in.Data<std::string>()[i]);
-    }
-  } else if (std::is_same<DstType, uint64_t>::value) {
-    auto* mutable_data = out.MutableData<uint64_t>();
-    for (int i = 0; i < len; ++i) {
-      mutable_data[i] = std::stoull(in.Data<std::string>()[i]);
-    }
-  } else {
-#ifdef ORT_NO_RTTI
-    ORT_THROW("Unsupported type in cast op");
-#else
-    ORT_THROW("Unsupported type in cast op: from String to ", typeid(DstType).name());
-#endif
+typename std::enable_if<std::is_integral<DstType>::value && std::is_unsigned<DstType>::value, void>::type
+CastFromString(const std::string& input, DstType& output) {
+  static_assert(sizeof(DstType) <= sizeof(unsigned long long),
```

Comment:
nit: would using uint64_t be clearer than 'unsigned long long'?

Reply:
using unsigned long long here because of the call to std::stoull

---
In reply to: [564884415](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564884415) [](ancestors = 564884415)